### PR TITLE
Would JMS info fit this project?

### PR DESCRIPTION
**Pull request by realModusOperandi**
This document has some nice config snippets, I've had some similar snippets for JMS in my notes that I wouldn't mind replacing with something like this. Is JMS among the info you'd like in the cheat sheet? (wasn't sure where to put it)

### CODE REVIEW

File: docs/index.adoc [hunk 1]
```diff
@@ -592,6 +592,115 @@ public class SecuredServiceEndpoint {
...
+
+.server.xml
+[source,xml]
+----
+<featureManager>
+    <feature>wasJmsServer-1.0</feature>
+</featureManager>
+
+<messagingEngine>
+    <queue id="libertyQueue" />
```
Comment (aguibert) [inline docs/index.adoc]: for the IDs in the config snippets, lets use IDs such as `myQueue` and `myTopic`, which I think helps convey that the values don't have any special meaning and that the reader can use whatever value they see fitting

File: docs/index.adoc [hunk 2]
```diff
@@ -592,6 +592,115 @@ public class SecuredServiceEndpoint {
 
 For more details see this guide: https://openliberty.io/guides/microprofile-jwt.html[Securing microservices with JWTs]
 
+== Messaging
+
+Applications can pass messages among themselves with the Java Message Service (JMS) APIs. Liberty can act as a server of JMS destinations (queues and topics,) or it can connect to another server to access those destinations--either another Liberty server or IBM MQ.
```
Comment (aguibert) [inline docs/index.adoc]: Could you add an extra sentence/paragraph explaining what queues and topics are and when to use one vs. the other? AFAIK a Queue is when each message is processed exactly once, and a Topic is more like a subscription model where zero-to-many processes could read each message.

Comment (aguibert) [reply docs/index.adoc]: Also, are the only other supported destination servers Liberty or IBM MQ? For example, does Kafka work?

File: docs/index.adoc [hunk 3]
```diff
@@ -592,6 +592,115 @@ public class SecuredServiceEndpoint {
...
+    <topicSpace id="libertyTopic" />
+</messagingEngine>
+----
+
+If you want applications on other servers to connect to your JMS server, create a `wasJmsEndpoint`. Without this, only applications served by the same Liberty server can acccess the JMS destinations.
+
+.server.xml
+[source,xml]
+----
+<wasJmsEndpoint id="InboundJmsCommsEndpoint"
```
Comment (aguibert) [inline docs/index.adoc]: is the ID significant here? If yes, you can leave as is. If not, lets change to `myEndpoint` or omit it entirely

File: docs/index.adoc [hunk 4]
```diff
@@ -592,6 +592,115 @@ public class SecuredServiceEndpoint {
...
+<featureManager>
+    <feature>wasJmsClient-2.0</feature>
+</featureManager>
+
+<jmsConnectionFactory jndiName="jms/localCF">
+    <properties.wasJms />
+</jmsConnectionFactory>
+
+<jmsConnectionFactory jndiName="jms/remoteCF">
+    <properties.wasJms remoteServerAddress="example.com:7276:BootstrapBasicMessaging" />
```
Comment (aguibert) [inline docs/index.adoc:652]: What does `BootstrapBasicMessaging` in the URL mean here?

File: docs/index.adoc [hunk 5]
```diff
@@ -592,6 +592,115 @@ public class SecuredServiceEndpoint {
...
+.server.xml
+[source,xml]
+----
+<wasJmsEndpoint id="InboundJmsCommsEndpoint" 
+                host="*" 
+                wasJmsPort="7276" 
+                wasJmsSSLPort="7286" />
+----
+
+=== JMS Client
```
Comment (aguibert) [inline docs/index.adoc:636]: This section has some good config snippets in it, but lets finish connecting the dots by (concisely) showing how an application might utilize this configuration. No need to have a 1-1 Java example for each config element, but maybe something that demonstrates how to send/receive a simple JMS message?

File: docs/index.adoc [hunk 6]
```diff
@@ -592,6 +592,115 @@ public class SecuredServiceEndpoint {
...
+----
+
+=== IBM MQ Client
+In addition to other Liberty servers, Liberty can connect to IBM MQ servers and allow applications to access destinations hosted in MQ Queue Managers. You don't need the `wasJmsClient` feature to connect to MQ.
+
+.server.xml
+[source,xml]
+----
+<resourceAdapter id="mqJmsRa" location="/path/to/wmq.jmsra.rar">
+    <classloader apiTypeVisibility="spec, ibm-api, api, third-party"/>
```
Comment (aguibert) [inline docs/index.adoc]: This can be more concisely written as:
```xml
<classloader apiTypeVisibility="+third-party"/>
```

File: docs/index.adoc [hunk 7]
```diff
@@ -592,6 +592,115 @@ public class SecuredServiceEndpoint {
...
+           -e MQ_ADMIN_PASSWORD=testpassword \
+           -p 1414:1414 \
+           -p 9443:9443 \
+           ibmcom/mq:9.1.4.0
+----
+
+The `wmq.jmsra.rar` file needed to connect to MQ can be found in the running Docker container using the command 
+
+[source, shell]
+docker cp mq-liberty:/opt/mqm/java/lib/jca/wmq.jmsra.rar /destination/folder
```
Comment (aguibert) [inline docs/index.adoc:719]: I didn't check but does this render properly without having the `----` lines before/after it?

**Review by aguibert — COMMENTED** [review]
Thanks @realModusOperandi! Yes a JMS section would be great for this document, I just have very little JMS experience so I didn't know what to add.

Added a few minor comments, but the main one is that I think we need some small Java code snippet to help the reader draw the link between server config and their application code.

**Comment by rieckpil** [comment]
After recording my latest YouTube video about Jakarta Messaging (https://www.youtube.com/watch?v=XhIX_tQgnng - publishing on Wednesday), I wanted to also contribute a JMS section to your cheat sheet and just saw this PR. If you want, you can include a link to the video as it uses Open Liberty and also shows how to activate a Message Driven Bean

**Comment by realModusOperandi** [comment]
I'll add a section on activation specs once I resolve these comments.

File: docs/index.adoc (hunk 2 above)
Comment (realModusOperandi) [reply docs/index.adoc]: I honestly don't know about others. They _should_ work generically if there exists an appropriate resource adapter similar to MQ does it but I didn't want to speak when I had no experience trying it out. I can throw in a sentence about other JMS-compatible messaging platforms.

File: docs/index.adoc (hunk 3 above)
Comment (realModusOperandi) [reply docs/index.adoc]: I'll probably just get rid of it for brevity's sake.

File: docs/index.adoc (hunk 4 above)
Comment (realModusOperandi) [reply docs/index.adoc:652]: I'll add a sentence describing the remoteServerAddress format, but basically it's kind of a holdover from tWAS that specifies a non-SSL connection. BootstrapSecureMessaging would specify an SSL connection.

File: docs/index.adoc (hunk 6 above)
Comment (realModusOperandi) [reply docs/index.adoc]: Interesting. I got that stanza from an official IBM doc so there may be a change they can make there.

File: docs/index.adoc (hunk 7 above)
Comment (realModusOperandi) [reply docs/index.adoc:719]: Yeah, in fact it kept rendering the dashes inside the snippet if I put them there. Probably has to do with only being one line.

File: docs/index.adoc (hunk 2 above)
Comment (aguibert) [reply docs/index.adoc]: That's fine if you're not certain about others. If you're not certain we don't need to name them explicitly, but we should at least allude to the possibility of other compatible messaging servers besides the the one listed (i.e. saying "a or b" vs saying "a, b, etc")

**Comment by realModusOperandi** [comment]
See how those look. I'll add a section on Activation Specs and MDBs when I have more time.

**Review by aguibert — APPROVED** [review]
looks good, thanks @realModusOperandi

**Pull request merged**